lib/app.js

```diff
@@ -137,53 +163,100 @@ Matrix.localization.init(Matrix.localesFolder, Matrix.config.locale, function ()
+              endIt();
+            } else if(status !='active'){
+              console.log(t('matrix.stop.stop_app_status_error') + ':', app);
+              endIt();
+            } else{
+              console.log(t('matrix.stop.stopping_app') + ': ', app);
+              if (options.all) {
+                //FIXME: hacky, passes this to server
+                app = 'all-applications'
+              }
+              Matrix.loader.start();
+
+              Matrix.firebase.app.watchStatus(appId, function (status) {
+                Matrix.loader.stop();
+                if(status =='inactive'){
+                  console.log(t('matrix.stop.stop_app_successfully') + ':', app);
+                  endIt();
+                }else{
+                  console.log(t('matrix.stop.stop_app_error') + ':', app);
+                }
+              });
+
+              Matrix.api.app.stop(app, Matrix.config.device.identifier, function (err) {
+                Matrix.loader.stop();
+                if (err) {
+                  console.log(t('matrix.stop.stop_app_error') + ':', app , ' (' + err.message.red + ')');
+                  endIt();
+                }
+              });
+            }
+          });
         });
-      }
+      });
     });
   //
   program
     .command('restart [app]')
     .description(t('matrix.help_restart'))
     .action(function (app) {
-      Matrix.validate.user(); //Make sure the user has logged in
-      Matrix.validate.device(); //Make sure the user has logged in
 
-      if (_.isUndefined(app)) {
-        console.log('\n> matrix restart ¬\n');
-        console.log('\t    matrix restart <app> -', t('matrix.restart.help', {app: '<app>'}).grey)
-        endIt();
-      } else {
-        console.log(t('matrix.restart.restarting_app') + ': ', app);
-        Matrix.loader.start();
-        Matrix.api.app.restart(app, function (err, res) {
-          Matrix.loader.stop();
-          if (!err) {
-            console.log(t('matrix.restart.restart_app_successfully') + ':', app, res);
-          } else {  
-            console.log(t('matrix.restart.restart_app_error') + ':', app , ' (' + err.message.red + ')');
-          } 
-          endIt();
+      //Make sure the user has logged in
+      Matrix.validate.user();
+      Matrix.validate.device();
+
+      Matrix.firebaseInit(function () {
+        Matrix.firebase.app.getIDForName( app, function(err, appId){
+          if (err) return console.error(err);
+          debug('appId>', appId);
+          Matrix.firebase.app.getStatus(appId, function (status) {
+            debug("Status Watch: " + Matrix.config.user.id + '>' + Matrix.config.device.identifier + '>' + appId + '>' + status);
+            if (_.isUndefined(app)) {
+              console.log('\n> matrix restart ¬\n');
+              console.log('\t    matrix restart <app> -', t('matrix.restart.help', {app: '<app>'}).grey)
+              endIt();
+            } else if(status!= 'active'){
```

Comment:
restart should only function if it's active

restart basically just does

stop then start on matrix os side